src/compiler/checker.ts

```diff
@@ -8095,7 +8106,7 @@ namespace ts {
                 }
                 const propName = propAccess.name.text;
                 const propType = getTypeOfPropertyOfType(type, propName);
-                if (!propType || !isStringLiteralUnionType(propType)) {
+                if (!propType || !(isStringLiteralUnionType(propType) || (type.flags & TypeFlags.UnionOrIntersection))) {

```

Comment:
You don't want to narrow if the original type was a union? That doesn't sound right.